cdap-ranger/cdap-ranger-lookup/src/main/java/co/cask/cdap/security/authorization/ranger/lookup/client/CDAPRangerLookupClient.java

```diff
@@ -119,6 +176,84 @@ private void initConnection() throws IOException {
+            try {
+              List<String> list = null;
+              if (resource.trim().equalsIgnoreCase(RangerCommons.KEY_NAMESPACE)) {
+                list = getNamespaces(resourceMap.get(RangerCommons.KEY_NAMESPACE));
+              } else {
+                // All of these belong to a namespace
+                Preconditions.checkNotNull(resourceMap.get(RangerCommons.KEY_NAMESPACE));
+                NamespaceId namespace = new NamespaceId(resourceMap.get(RangerCommons.KEY_NAMESPACE).get(0));
+                if (resource.trim().equalsIgnoreCase(RangerCommons.KEY_STREAM)) {
+                  list = getStreams(resourceMap.get(RangerCommons.KEY_NAMESPACE), namespace);
+                } else if (resource.trim().equalsIgnoreCase(RangerCommons.KEY_APPLICATION)) {
+                  list = getApplications(resourceMap.get(RangerCommons.KEY_APPLICATION), namespace);
+                } else if (resource.trim().equalsIgnoreCase(RangerCommons.KEY_DATASET)) {
+                  list = getDatasets(resourceMap.get(RangerCommons.KEY_DATASET), namespace);
+                } else if (resource.trim().equalsIgnoreCase(RangerCommons.KEY_PROGRAM)) {
+                  Preconditions.checkNotNull(resourceMap.get(RangerCommons.KEY_APPLICATION));
+                  ApplicationId applicationId = new ApplicationId(namespace.getNamespace(), resourceMap.get
+                    (RangerCommons.KEY_APPLICATION).get(0));
+                  list = getPrograms(resourceMap.get(RangerCommons.KEY_PROGRAM), applicationId);
+                } else if (resource.trim().equalsIgnoreCase(RangerCommons.KEY_ARTIFACT)) {
+                  list = getArtifacts(resourceMap.get(RangerCommons.KEY_ARTIFACT), namespace);
+                } else if (resource.trim().equalsIgnoreCase(RangerCommons.KEY_DATASET_MODULE)) {
+                  list = getDatasetModules(resourceMap.get(RangerCommons.KEY_DATASET_MODULE), namespace);
+                } else if (resource.trim().equalsIgnoreCase(RangerCommons.KEY_DATASET_TYPE)) {
+                  list = getDatasetTypes(resourceMap.get(RangerCommons.KEY_DATASET_TYPE), namespace);
+                } else if (resource.trim().equalsIgnoreCase(RangerCommons.KEY_SECUREKEY)) {
+                  list = getSecureKeys(resourceMap.get(RangerCommons.KEY_SECUREKEY), namespace);
+                }
+              }
+              Preconditions.checkNotNull(list, "Failed to list resources of type %s", resource.trim());
+              if (!userInput.isEmpty()) {
+                for (String value : list) {
+                  if (value.startsWith(userInput)) {
+                    retList.add(value);
+                  }
+                }
+              } else {
+                retList.addAll(list);
+              }
+            } catch (Exception ex) {
+              LOG.error("Error getting resource.", ex);
+            }
+            return retList;
+          }
+        };
+
+        synchronized (this) {
```

Comment:
That's what I'm trying to understand - why we need result set to be created in a synchronized block. All our clients are thread-safe, right? In that case I don't think the synchronized block is needed. What do you say?